os/kernel/binary_manager/binary_manager_recovery.c

```diff
@@ -91,59 +97,94 @@ static void recovery_release_binary_sem(int binid)
+static void recovery_recover_tcb(struct tcb_s *tcb)
 {
-	int binid;
-	irqstate_t flags;
-
-	binid = (int)arg;
-	if (binid < 0) {
-		return;
-	}
+	sem_t *sem;
+	int state;
 
-	if (tcb->group->tg_loadtask == binid) {
+	state = tcb->task_state;
 
-		/* Recover semaphores, message queue, and watchdog timer resources.*/
-		task_recover(tcb);
+	if (tcb->waitdog) {
+		(void)wd_delete(tcb->waitdog);
+		tcb->waitdog = NULL;
+	}
 
-		flags = irqsave();
-		/* Remove the TCB from the task list associated with the state */
-		dq_rem((FAR dq_entry_t *)tcb, (dq_queue_t *)g_tasklisttable[tcb->task_state].list);
-		sched_addblocked(tcb, TSTATE_TASK_INACTIVE);
-		irqrestore(flags);
-		bmllvdbg("Remove pid %d from task list\n", tcb->pid);
+	if (state == TSTATE_WAIT_SEM) {
+		sem = tcb->waitsem;
+		DEBUGASSERT(sem != NULL && sem->semcount < 0);
+		sem_canceled(tcb, sem);
+		sem->semcount++;
+		tcb->waitsem = NULL;
+	} else if (state == TSTATE_WAIT_MQNOTEMPTY) {
+		DEBUGASSERT(tcb->msgwaitq && tcb->msgwaitq->nwaitnotempty > 0);
+		tcb->msgwaitq->nwaitnotempty--;
+	} else if (state == TSTATE_WAIT_MQNOTFULL) {
+		DEBUGASSERT(tcb->msgwaitq && tcb->msgwaitq->nwaitnotfull > 0);
+		tcb->msgwaitq->nwaitnotfull--;
 	}
 }
 
-/****************************************************************************
- * Name: recovery_exclude_scheduling
- *
- * Description:
- *   This function will move all the tasks and threads created by the binary
- *   i.e input pid to inactive task list to exclude them from scheduling.
- *
- * Input parameters:
- *   pid   -   The pid of the binary, whoes all children to be killed
- *
- * Returned Value:
- *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
- *
- ****************************************************************************/
-static int recovery_exclude_scheduling(int binid)
+static int recovery_exclude_binary(pid_t binid)
 {
-	if (binid < 0) {
+	irqstate_t flags;
+	struct tcb_s *tcb;
+
+	tcb = sched_gettcb(binid);
+	if (tcb == NULL) {
 		return ERROR;
 	}
 
-	/* Exclude all tasks and pthreads created in a binary which has 'binid' from scheduling */
-	sched_foreach(recovery_exclude_scheduling_each, (FAR void *)binid);
+	/* A main task which has 'binid' is a head of binary list */
+	while (tcb) {
+		if (tcb->task_state != TSTATE_TASK_INACTIVE) {
+			flags = irqsave();
+			/* Recover semaphores, message queue, and watchdog timer resources.*/
+			recovery_recover_tcb(tcb);
+
+			/* Remove the TCB from the task list associated with the state */
+			BM_EXCLUDE_SCHEDULING(tcb);
+			tcb = tcb->bin_flink;
+			irqrestore(flags);
+		}
+	}
 
 	/* Release all semaphores held by the threads in binary */
 	recovery_release_binary_sem(binid);
 
 	return OK;
 }
 
+/* Exclude real-time threads in a same binary from scheduling */
+void recovery_exclude_rtthreads(struct tcb_s *tcb)
+{
+	struct tcb_s *prev;
+	struct tcb_s *next;
+
+	prev = tcb->bin_blink;
+	next = tcb->bin_flink;
+
+	while (prev) {
```

Comment:
How about merging while condition and if condition?

Reply:
It can't be merged because of blink.